docs/configs/janusgraph-cfg.md

```diff
@@ -406,6 +405,17 @@ CQL storage backend options
 | storage.cql.use-external-locking | True to prevent JanusGraph from using its own locking mechanism. Setting this to true eliminates redundant checks when using an external locking mechanism outside of JanusGraph. Be aware that when use-external-locking is set to true, that failure to employ a locking algorithm which locks all columns that participate in a transaction upfront and unlocks them when the transaction ends, will result in a 'read uncommitted' transaction isolation level guarantee. If set to true without an appropriate external locking mechanism in place side effects such as dirty/non-repeatable/phantom reads should be expected. | Boolean | false | MASKABLE |
 | storage.cql.write-consistency-level | The consistency level of write operations against Cassandra | String | QUORUM | MASKABLE |
 
+### storage.cql.executor-service
+Configuration options for CQL executor service which is used to process CQL queries.
+
+
+| Name | Description | Datatype | Default Value | Mutability |
+| ---- | ---- | ---- | ---- | ---- |
+| storage.cql.executor-service.class | The implementation of `ExecutorService` to use. The class full name of the class which extends `ExecutorService` which has either a public constructor with `ExecutorServiceConfiguration` argument (preferred constructor) or a public parameterless constructor. Other accepted options are: `fixed` - fixed thread pool size of `max-pool-size` size; `cached` - cached thread pool size; | String | fixed | LOCAL |
```

Comment:
[DISCUSS] I am thinking if we should allow not using executor service at all. Maybe I will try what happens when we don't use a thread pool and vavr at all, and then compare the performance.

Reply:
I actually was thinking about it as well. I'm not sure why we actually need CQL thread pool at all. It looks like with current implementation in `master` branch we can execute up to 10 parallel CQL queries and never more because this thread pool is limiting us which is crazy unless I'm missing something. Maybe it was introduced to limit the amount of parallel interaction with the same CQL storage backend but I'm not sure. Nevertheless, with this PR users can either increase this thread pool size or just simply switch to cached thread pool size.
In general, I think we need to allow users to optionally disable / enable CQL thread pool usage as we do with parallel backend queries thread pool. That said, I think this functionality should go to a separate PR. Would be great if you could check performance difference between using and not using thread pool there.

Reply:
Checking `CQLStoreManager.mutateManyUnlogged` closer now I see that we may execute a sequence of requests in parallel. Thus, having the executor service make sense there. At the end we would still wait until all the requests are finished but with executor service we gain 2 features here:
- executing a sequence of requests in parallel
- limiting of the amount of parallel executions of such sequences

I assume executing CQL requests sequentially without the executor service will make queries always take only 1 current thread which might be slower but won't block other queries.
With limited executor service we can process such queries in parallel but a single query can potentially take all the available threads which essentially will delay the execution of other queries.
With unlimited executor service (i.e. `cached` executor service OR custom unlimited executor service) we will execute all sequences of all queries in parallel but we may potentially acquire too many threads which may lead to OOM or decrease the overall performance due to threads context switching overhead.

I think we should use CQL executor service by default (as we do right now) but make it configurable (as this PR does). Executing queries without executor service is a separate feature which in some situations I think could be beneficial (resources or performance optimization) but we should double check if we gain any improvements when executing queries not via executor service.

Reply:
Agree that "Executing queries without executor service is a separate feature" that should be done in a separate PR.

I suspect the following statement is not accurate:
> executing CQL requests sequentially without the executor service will make queries always take only 1 current thread

The reason is `this.session.executeAsync` is asynchronous and returns a `CompletionStage`. When we have multiple batches, they are not executing in sequence. We only need to wait for their completeness in sequence.

Reply:
Oh, you are right. I was thinking if we change executeAsync to execute then we would have sequential queries execution. If we leave executeAsync then all queries will be executed in parallel but then we won't be able to limit the amount of parallel queries if we don't have executorService.

Reply:
> we won't be able to limit the amount of parallel queries if we don't have executorService

That's my guess as well: https://github.com/JanusGraph/janusgraph/discussions/2545#discussioncomment-620084

Reply:
Sorry if I'm missing something but if the "CQL executor service" was removed, wouldn't the Backend threadpool size limit the number of concurrent requests anyway?

Reply:
> wouldn't the Backend threadpool size limit the number of concurrent requests anyway

That's a good point, but it is not always the case.

https://github.com/JanusGraph/janusgraph/blob/a28e314dcf09d1ae2f6b2ab341fe3f7f44fb5cc4/janusgraph-core/src/main/java/org/janusgraph/diskstorage/BackendTransaction.java#L301-L304

From above we can see that backend thread pool is not used when 1) PARALLEL_BACKEND_OPS is disabled, OR 2) the number of queries is one.

I would say the behavior of the combination usage of the backend thread pool and CQL thread pool is very confusing and it would be great if we can document it.

Reply:
Limiting the number of concurrent CQL queries seems not very useful to me anyways, because users can tune the CQL driver configurations to limit using configurations like `MAX_REQUESTS_PER_CONNECTION`.

Reply:
I think @li-boxuan is right here. I added a separate ticket #2707 to track the feature request to add possibility to disable ExecutorService for CQL requests.